keps/sig-auth/20200507-external-tls-certificate-authenticator/README.md

```diff
@@ -0,0 +1,749 @@
+- External plugin prompts the user for a PIN to perform signing operation (if
+  the PIN is not provided in the configuration file).
+- External plugin returns a client certificate and a signature to client-go via
+  the authentication provider.
+- API server verifies the signature and processes the request.
+
+<!-- - Credential plugin prompts the user for LDAP credentials, exchanges credentials with external service for a token.
+- Credential plugin returns token to client-go, which uses it as a bearer token against the API server.
+- API server uses the webhook token authenticator to submit a TokenReview to the external service.
+- External service verifies the signature on the token and returns the user’s username and groups. -->
+
+<!-- #### Story 2 -->
+
+### Notes/Constraints/Caveats
+
+<!--
+What are the caveats to the proposal? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+-->
+
+ It was requested by the sig-auth community to come up with a solution that does
+ not require CGO during the build process and preferably does not add new
+ dependencies to kubectl.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly. For
+example, consider both security and how this will impact the larger kubernetes
+ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your change
+are understandable.  This may include API specs (though not always required) or
+even code snippets.  If there's any ambiguity about HOW your proposal will be
+implemented, this is the place to discuss them.
+-->
+
+The complete solution for external TLS certificate signing includes two
+components:
+
+- an authentication provider within kubectl/client-go, and
+- an external plugin, which provides the actual certificate and signing
+  services.
+
+![UML sequence diagram](uml_sequence_diagram.png)
+
+The main role of the authentication provider is to pass along requests for
+obtaining a TLS certificate and signing to the external plugin, retrieve the
+responses and return them to the initial caller.
+
+Users would be required to install an external plugin (possibly along with
+dependencies) on their workstation.
+
+### Configuration
+
+Configuration of the both components (authentication provider and external
+plugin) is specified in the kubeconfig files as part of the user fields. The
+authentication provider is responsible for reading the configuration from the
+kubeconfig file and exposing the configuration parameters to the external
+plugin.
+
+The internal authentication provider requires only a single parameter, the path
+to the external plugin `pathExec`.
```

Comment:
off topic, but just to note that the method of executing arbitrary binaries from a kubeconfig is already considered insecure by some bloggers too:
https://banzaicloud.com/blog/kubeconfig-security/